packages/gateway/scripts/denylist.js

```diff
@@ -0,0 +1,149 @@
+import fetch from '@web-std/fetch'
+import path from 'path'
+import toml from 'toml'
+import { CID } from 'multiformats'
+import { base32 } from 'multiformats/bases/base32'
+import { sha256 } from 'multiformats/hashes/sha2'
+import fs from 'fs'
+import * as uint8arrays from 'uint8arrays'
+
+/**
+ * @typedef {{ id: string, title: string }} Namespace
+ * @typedef {{ name: string, metadata: any }} Key
+ * @typedef {{ key: string, value: any, metadata?: any }} BulkWritePair
+ */
+
+const rootDir = path.dirname(path.dirname(import.meta.url))
+const wranglerConfigPath = `${rootDir}/wrangler.toml`
+const denyListPath = `${rootDir}/denylist.json`
+
+const DENY_LIST_SOURCES = [
+  'https://badbits.dwebops.pub/denylist.json',
+  denyListPath,
+]
+
+export async function denylistSyncCmd({ env }) {
+  const cfApiToken = mustGetEnv('CF_API_TOKEN')
+
+  const wranglerConfig = await getWranglerToml(wranglerConfigPath)
+  const wranglerEnvConfig = wranglerConfig.env[env]
+  if (!wranglerEnvConfig) {
+    throw new Error(`missing wrangler configuration for env: ${env}`)
+  }
+  console.log(`🧩 using wrangler config: ${wranglerConfigPath}`)
+
+  const cfAccountId = wranglerEnvConfig.account_id
+  if (!cfAccountId) {
+    throw new Error(`missing Cloudflare account_id in env: ${env}`)
+  }
+  console.log(`🏕 using env: ${env} (${cfAccountId})`)
+
+  const kvNamespaces = wranglerEnvConfig.kv_namespaces || []
+  const denyListKv = kvNamespaces.find((kv) => kv.binding === 'DENYLIST')
+  if (!denyListKv) {
+    throw new Error('missing binding in kv_namespaces: DENYLIST')
+  }
+  console.log(`🪢 using KV binding: DENYLIST (${denyListKv.id})`)
+
+  for (const url of DENY_LIST_SOURCES) {
+    console.log(`🦴 fetching ${url}`)
+    const denyList = await getDenyList(url)
+    const kvs = denyList.map(({ anchor: key, status, reason }) => ({
+      key,
+      value: { status, reason },
```

Comment:
given we won't have status/reason in any place at the moment, wondering if we should store an origin property with where the CID+Path combo is registered, or add a default value to reason including it? what do you think?

Reply:
I'd probably not have a default reason that includes the origin because in the case when we need to set a reason it would not be a reliable way to know the origin.

I'd be ok with putting `origin` as a separate property...but where would we use it?